### PR TITLE
Type Ignore Fixes Part Four

### PR DESCRIPTION
**Pull request by GavinZZ**
### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ConnorRobertson — APPROVED** [review]
LGTM

File: samtranslator/model/s3_utils/uri_parser.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def construct_image_code_object(image_uri, logical_id, property_name):  # type:
     return {"ImageUri": image_uri}
 
 
-def construct_s3_location_object(location_uri, logical_id, property_name):  # type: ignore[no-untyped-def]
+def construct_s3_location_object(location_uri: Any, logical_id: str, property_name: str) -> Dict[str, Any]:
```
Comment (aaythapa) [inline samtranslator/model/s3_utils/uri_parser.py:65]: wouldn't `location_uri` be `Union[str, Dict[str, Any]]`

File: samtranslator/model/lambda_.py [hunk 2]
```diff
@@ -152,6 +152,13 @@ class LambdaLayerVersion(Resource):
         "LicenseInfo": PropertyType(False, IS_STR),
     }
 
+    Content: Dict[str, Any]
+    Description: Optional[Intrinsicable[str]]
+    LayerName: Optional[Intrinsicable[str]]
+    CompatibleArchitectures: Optional[List[Union[str, Dict[str, Any]]]]
+    CompatibleRuntimes: Optional[List[Union[str, Dict[str, Any]]]]
+    LicenseInfo: Optional[Intrinsicable[str]]
```
Comment (aahung) [inline samtranslator/model/lambda_.py:160]: nit: any of these can be replaced with `PassThrough`?

File: samtranslator/model/sam_resources.py [hunk 3]
```diff
@@ -694,7 +702,7 @@ def _validate_dlq(self, dead_letter_queue: Dict[str, Any]) -> None:
                 self.logical_id, "'DeadLetterQueue' requires Type of {}".format(valid_dlq_types)
             )
 
-    def _event_resources_to_link(self, resources):  # type: ignore[no-untyped-def]
+    def _event_resources_to_link(self, resources: Any) -> Dict[str, Any]:
```
Comment (aaythapa) [inline samtranslator/model/sam_resources.py:705]: Isn't `resources` always a Dict?

File: samtranslator/model/lambda_.py (hunk 2 above)
Comment (GavinZZ) [reply samtranslator/model/lambda_.py:160]: Good call.

Comment (GavinZZ) [reply samtranslator/model/lambda_.py:160]: Actually no, this is `AWS::Lambda::LayerVersion`. None of the properties should be PassThrough. `PassThrough` should only be used on SAM resources right? @aahung

File: samtranslator/model/sam_resources.py [hunk 4]
```diff
@@ -722,9 +730,14 @@ def order_events(event):  # type: ignore[no-untyped-def]
...
         return event_dict.get("Properties", {}).get("Path", logical_id)
 
-    def _generate_event_resources(  # type: ignore[no-untyped-def]
-        self, lambda_function, execution_role, event_resources, intrinsics_resolver, lambda_alias=None
-    ):
+    def _generate_event_resources(
+        self,
+        lambda_function: LambdaFunction,
+        execution_role: Optional[IAMRole],
+        event_resources: Any,
```
Comment (aaythapa) [inline samtranslator/model/sam_resources.py:737]: This will be `Dict[Any, Any]` too I think

**Review by aaythapa — APPROVED** [review]

**Review by aahung — APPROVED** [review]

**Review by xazhao — APPROVED** [review]

**Pull request merged**